modifiers/godot/voxel_modifier_gd.cpp

```diff
@@ -115,24 +85,63 @@ void VoxelModifier::_notification(int p_what) {
 				_modifier_id = 0;
 			}
 		} break;
+		case Node3D::NOTIFICATION_TRANSFORM_CHANGED: {
+			if (_volume == nullptr) {
+				Node *parent = get_parent();
+				VoxelLodTerrain *volume = Object::cast_to<VoxelLodTerrain>(parent);
+				// Using get_parent_node_3d when terrain is immediate parent
+				// would be a breaking change as it returns null if node is toplevel
+				if (volume == nullptr) {
+					Node3D *grandparent = get_parent_node_3d();
+					volume = Object::cast_to<VoxelLodTerrain>(grandparent);
+					while (grandparent != nullptr && volume == nullptr) {				
+						grandparent = grandparent->get_parent_node_3d();
+						volume = Object::cast_to<VoxelLodTerrain>(grandparent);
+					}
+				}
+				_volume = volume;
+			
+
+				if (_volume != nullptr) {
+					VoxelData &voxel_data = _volume->get_storage();
+					VoxelModifierStack &modifiers = voxel_data.get_modifiers();
+					const uint32_t id = modifiers.allocate_id();
+					zylann::voxel::VoxelModifier *modifier = create(modifiers, id);
+
+					if (modifier->is_sdf()) {
+						zylann::voxel::VoxelModifierSdf *sdf_modifier =
+								static_cast<zylann::voxel::VoxelModifierSdf *>(modifier);
+						sdf_modifier->set_operation(to_op(_operation));
+						sdf_modifier->set_smoothness(_smoothness);
+					}
+					// This transform is wrong, but will be updated with NOTIFICATION_TRANSFORM_CHANGED
+					// After node enters the tree
+					modifier->set_transform(get_transform());
+					_modifier_id = id;
+					// TODO Optimize: on loading of a scene, this could be very bad for performance because there could be,
+					// a lot of modifiers on the map, but there is no distinction possible in Godot at the moment...
+					post_edit_modifier(*_volume, modifier->get_aabb());
+				}
+			}
 
-		case Node3D::NOTIFICATION_LOCAL_TRANSFORM_CHANGED: {
+			update_configuration_warnings();
 			if (_volume != nullptr && is_inside_tree()) {
 				VoxelData &voxel_data = _volume->get_storage();
 				VoxelModifierStack &modifiers = voxel_data.get_modifiers();
 				zylann::voxel::VoxelModifier *modifier = modifiers.get_modifier(_modifier_id);
 				ZN_ASSERT_RETURN(modifier != nullptr);
 
 				const AABB prev_aabb = modifier->get_aabb();
-				modifier->set_transform(get_transform());
+				Transform3D terrain_local = _volume->get_global_transform().affine_inverse() * get_global_transform();
+				// If the terrain moved, the modifiers do not need to be updated.
+				if (terrain_local == modifier->get_transform()) {
```

Comment:
I thought of this too, unfortunately at this point it already required computing a bunch of transforms, which can also suffer from floating point precision issues so the check might still fail

Reply:
I observed some floating point errors causing unnecessary updates. Do you think it would work to check whether the transforms differ by some small amount instead of checking for equality?

Reply:
I would rather prefer nothing to happen at all if possible... but it seems like supporting nesting requires extra complexity/workarounds I didnt want to add originally (especially if direct children have to pay the same cost). I don't know how else it can be solved. I'm also not aware of nodes in Godot that have transform-dependent logic getting register all the way to a specific Node3D parent like this.